rfc/rfc-56/rfc-56.md

```diff
@@ -0,0 +1,226 @@
+
+## Abstract
+
+As you scale your Apache Hudi workloads over Cloud object stores like Amazon S3, there is potential of hitting request
+throttling limits which in-turn impacts performance. In this RFC, we are proposing to support an alternate storage
+layout that is optimized for Amazon S3 and other cloud object stores, which helps achieve maximum throughput and
+significantly reduce throttling.
+
+In addition, we are proposing an interface that would allow users to implement their own custom strategy to allow them
+to distribute the data files across cloud stores, hdfs or on prem based on their specific use-cases.
+
+## Background
+
+Apache Hudi follows the traditional Hive storage layout while writing files on storage:
+- Partitioned Tables: The files are distributed across multiple physical partition folders, under the table's base path.
+- Non Partitioned Tables: The files are stored directly under the table's base path.
+
+While this storage layout scales well for HDFS, it increases the probability of hitting request throttle limits when
+working with cloud object stores like Amazon S3 and others. This is because Amazon S3 and other cloud stores [throttle
+requests based on object prefix](https://aws.amazon.com/premiumsupport/knowledge-center/s3-request-limit-avoid-throttling/).
+Amazon S3 does scale based on request patterns for different prefixes and adds internal partitions (with their own request limits),
+but there can be a 30 - 60 minute wait time before new partitions are created. Thus, all files/objects stored under the
+same table path prefix could result in these request limits being hit for the table prefix, specially as workloads
+scale, and there are several thousands of files being written/updated concurrently. This hurts performance due to
+re-trying of failed requests affecting throughput, and result in occasional failures if the retries are not able to
+succeed either and continue to be throttled.
+
+The traditional storage layout also tightly couples the partitions as folders under the table path. However,
+some users want flexibility to be able to distribute files/partitions under multiple different paths across cloud stores,
+hdfs etc. based on their specific needs. For example, customers have use cases to distribute files for each partition under
+a separate S3 bucket with its individual encryption key. It is not possible to implement such use-cases with Hudi currently.
+
+The high level proposal here is to introduce a new storage layout strategy, where all files are distributed evenly across
+multiple randomly generated prefixes under the Amazon S3 bucket, instead of being stored under a common table path/prefix.
+This would help distribute the requests evenly across different prefixes, resulting in Amazon S3 to create partitions for
+the prefixes each with its own request limit. This significantly reduces the possibility of hitting the request limit
+for a specific prefix/partition.
```

Comment:
Should we mention about dynamic sub-partitioning strategies in case one of these partitions crosses a threshold? - so that in case we need to implement such a scenario, the design is generic enough to handle it.

Reply:
+1 on dynamic prefix-splitting or rebalancing